### ISSUE TITLE
pc_expire_check broken (incorrect timer)

### ISSUE THREAD

**Issue by anacondaqq** [rathena/rathena#2438]
### Technical Information
rAthena hash: 5c7c0893203a6841ca2ee7c7743f56ccdff4042c
Client Ver: 2015-11-04aRagexe
Mode: pre-re / renewal (does not matter)

### Problem
pc.c: please see function below:
```cpp
void pc_expire_check(struct map_session_data *sd) {  
	/* ongoing timer */
	if( sd->expiration_tid != INVALID_TIMER )
		return;

	/* not within the next 24h, enable the global check */
	if( sd->expiration_time > (time(NULL) + ((60 * 60) * 24)) ) {

		/* global check not running, enable */
		if( pc_expiration_tid == INVALID_TIMER ) /* Starts in 1h, repeats every hour */
			pc_expiration_tid = add_timer_interval(gettick() + ((1000 * 60) * 60), pc_global_expiration_timer, 0, 0, ((1000 * 60) * 60));

		return;
	}

	sd->expiration_tid = add_timer(gettick() + (unsigned int)(sd->expiration_time - time(NULL)) * 1000, pc_expiration_timer, sd->bl.id, 0);
}
```

After script **.cpp** upgrade the line
`sd->expiration_tid = add_timer(gettick() + (unsigned int)(sd->expiration_time - time(NULL)) * 1000, pc_expiration_timer, sd->bl.id, 0);` now broken (incorrect timer, never triggered pc_expiration_timer)
Befre scrip **.cpp** upgrade everything was works fine.

### What need todo?
Please repair 
```cpp
sd->expiration_tid = add_timer(gettick() + (unsigned int)(sd->expiration_time - time(NULL)) * 1000, pc_expiration_timer, sd->bl.id, 0);
```
for setting a timer to kicking off a player from the game if timer expired. 

### How reproduce?
Its hard to explain, because there are several bugs together (related to rental items with more or less 24 hours expire timer), and totally different issues and scenarios, there is no sense at all to explain it, you can check that line above broken manually and it will be much faster to fix, than sitting and setting different vars. But if it required, just let me know, and I will post a full step by step guide how to reproduce the issue. Sorry for so short description of the issue.

**Comment by aleos89** [comment]
I created an expiration timer for my account, logged in, waited for the time and was properly kicked off. If you have any other ways to reproduce your error let me know, otherwise I can't reproduce this on d9a2bdb.

**Comment by cydh** [comment]
* [ ] rental time is less than 24 hours?
* [ ] after player get the rental item, the player were logged out?
* [ ] player that has no permanent SC/cart/falcon/mount?

**Comment by anacondaqq** [comment]
Just trigger the line above. It's hard for me to explain because I have found this bug in very rare and strange custom scenario. 

code what needs to execute in any custom scenario (does not matter which one, it will show you what I wanna to say, but cant 😭 

```c
sd->expiration_tid = add_timer(gettick() + (unsigned int)(sd->expiration_time - time(NULL)) * 1000, pc_expiration_timer, sd->bl.id, 0);
```

It must create a timer which will trigger after 1-second function below. 
But for some reason, the code line does not trigger timer function below.
And ther is a bug about what I'm trying to say =)

In other words expected next result:
You adding a timer, and when its expired, char must be kicked off from the game, but its never happened, because something is wrong with the line above (before cpp changes it was work).

```c
int pc_expiration_timer(int tid, unsigned int tick, int id, intptr_t data) {
	struct map_session_data *sd = map_id2sd(id);

	if( !sd ) return 0;

	sd->expiration_tid = INVALID_TIMER;

	if( sd->fd )
		clif_authfail_fd(sd->fd,10);

	map_quit(sd);

	return 0;
}
```

**Comment by Jeybla** [comment]
Isn't the expiration timer only used for account expiration for limited test accounts?
I tested it by setting the `expiration_time` in the database for the specific account and it works fine.
@anacondaqq can you please just give an example where you encountered this error. I'm quite sure, we're looking into the wrong code.

**Comment by lighta** [comment]
@anacondaqq 
idk if you have trace it but my guess is that "(sd->expiration_time - time(NULL)". was in your case returning a negative value.
Now since this is follow with a cast. (unsigned int). This will add a good amount of second, thus the (never call function).
Can you try add a abs to this. => abs(sd->expiration_time - time(NULL)). or add a breakpoint to see at which tick it was suppose to fire.

**Comment by aleos89** [comment]
Going to close this since none of the devs could reproduce it. If this is still occurring then we can reopen this.
